ICP-2367 Yale Locks send lock operations before notification events

greens opened the pull request:
We need to enforce the order of notification and lock in order to get the most information possible for the lock/unlock events. This attempts to delay the lock command event until after the notification event. Because of isStateChange, the lock event will actually essentially get ignored as long as the notification event comes.

greens committed:
ICP-2367 Yale Locks send lock operations before notification events

We need to enforce the order of notification and lock in order to get the most information possible for the lock/unlock events. This attempts to delay the lock command event until after the notification event. Because if isStateChange, the lock event will actually essentially get ignored as long as the notification event comes.

vaish-rahul commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -313,6 +313,17 @@ def zwaveEvent(DoorLockOperationReport cmd) {
 	result ? [createEvent(map), *result] : createEvent(map)
+	if (generatesDoorLockOperationReportBeforeAlarmReport()) {
+		// we're expecting lock events to come after notification events, but for specific yale locks they come out of order
+		runIn(3, "delayLockEvent", [data: [event: result]])
```
This won't work. I see a problem.

`result` will not have any data for locked/unlocked event as you are not putting that into the `result` list. So the changes will not work when expected to work. Besides this, the changes will also break the existing functionality of `refresh` in DTH.

greens replied:
Ah thanks, didn't notice there wasn't actually an assignment on the line above.

greens committed:
ICP-2367 Yale Locks send lock operations before notification events

greens commented:
@vaish-rahul This should fix it. Thanks for catching that. Refresh should still work, there will just be a bit more of a delay for Yale locks.

mckeed commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -312,7 +312,18 @@ def zwaveEvent(DoorLockOperationReport cmd) {
 		}
 	}
-	result ? [createEvent(map), *result] : createEvent(map)
+	result = result ? [createEvent(map), *result] : createEvent(map)
```
`createEvent` sets `isStateChange`. It would be better to move this line to the `else` block below and just send `map` to `delayLockEvent`.

vaish-rahul replied:
Good catch @mckeed

greens committed:
ICP-2367 Yale locks send notifications in unexpected order

This should delay the calling of createEvent on the doorLockOperation data so that the isStateChange flag should be set how we expect it to be.

vaish-rahul commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -312,18 +312,17 @@ def zwaveEvent(DoorLockOperationReport cmd) {
 
 def delayLockEvent(data) {
-	sendEvent(data.event)
+	sendEvent(data.result ? [createEvent(data.map), *data.result] : createEvent(data.map))
```
Now theoretically this looks correct. But today I tried this out by self publishing the DTH and `delayLockEvent` doesn't get called. I made `generatesDoorLockOperationReportBeforeAlarmReport` return true explicitly and `delayLockEvent` gets scheduled to be called in 3 seconds but it doesn't get called and as a result event doesn't get generated. May be I am missing something or something is broken in production server because of which `delayLockEvent` is not getting called. Either way, @greens can you try executing this in your account by self publishing and check whether `delayLockEvent` gets called?

greens replied:
I tested on my account with logging set up with the correct device and delayLockEvent was being called for me. 

Try testing it in the IDE with some additional logging statements. Because the event _shouldn't_ show up.

vaish-rahul replied:
I added a log in `delayLockEvent` and it didn't print. Anyway, if it's working for you then it's Ok.

greens committed:
ICP-2367 Yale locks send notifications in unexpected order.

Cleaning up some of the code and adding a debug log for this rare event.

vaish-rahul commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -314,15 +314,16 @@ def zwaveEvent(DoorLockOperationReport cmd) {
 def delayLockEvent(data) {
-	sendEvent(data.result ? [createEvent(data.map), *data.result] : createEvent(data.map))
+	log.debug "Sending cached lock operation: $data.map"
+	sendEvent(data.map)
```
Cool. I missed to check that `sendEvent` takes a Map and not a list. However, with this we will miss to send the `HubActions` added to `result` list in `zwaveEvent(DoorLockOperationReport cmd)`. Are we sure that this will not break anything?

mckeed replied:
Those hubActions are primarily for fixing an association issue that came up with first-generation Schlage locks. I expect it won't be a problem to skip them for recent Yale models.

greens committed:
Merge branch 'master' into ICP-2367

greens committed:
ICP-2367

Merge was bad.
